File: tests/python/test_fretboard_svg.py
"""Tests for the FretboardSVG class."""

import os
import tempfile

import pytest
from gingo import (
    Fretboard, FretboardSVG, Chord, Scale, ScaleType, Note, Field, Layout,
    Orientation, Handedness,
)


# ---------------------------------------------------------------------------
# Chord diagram
# ---------------------------------------------------------------------------

class TestFretboardSVGChord:
    def test_generates_valid_svg(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"))
        assert "<svg" in svg
        assert "</svg>" in svg
        assert 'xmlns="http://www.w3.org/2000/svg"' in svg

    def test_contains_title(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("Am7"))
        assert "Am7" in svg

    def test_has_style_block(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"))
        assert "<style>" in svg
        assert "</style>" in svg

    def test_has_string_lines(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"))
        assert 'class="str"' in svg

    def test_has_fret_lines(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"))
        assert 'class="frt"' in svg

    def test_has_dots(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"))
        assert 'class="dot"' in svg

    def test_on_cavaquinho(self):
        c = Fretboard.cavaquinho()
        svg = FretboardSVG.chord(c, Chord("CM"))
        assert "<svg" in svg
        assert "CM" in svg

    def test_on_bandolim(self):
        b = Fretboard.bandolim()
        svg = FretboardSVG.chord(b, Chord("CM"))
        assert "<svg" in svg


# ---------------------------------------------------------------------------
# Fingering
# ---------------------------------------------------------------------------

class TestFretboardSVGFingering:
    def test_renders_fingering(self):
        g = Fretboard.violao()
        f = g.fingering(Chord("Am"))
        svg = FretboardSVG.fingering(g, f)
        assert "<svg" in svg
        assert "Am" in svg


# ---------------------------------------------------------------------------
# Scale
# ---------------------------------------------------------------------------

class TestFretboardSVGScale:
    def test_generates_valid_svg(self):
        g = Fretboard.violao()
        svg = FretboardSVG.scale(g, Scale("C", ScaleType.Major), 0, 12)
        assert "<svg" in svg
        assert "</svg>" in svg

    def test_has_string_labels(self):
        g = Fretboard.violao()
        svg = FretboardSVG.scale(g, Scale("C", ScaleType.Major), 0, 5)
        assert "E4" in svg


# ---------------------------------------------------------------------------
# Note
# ---------------------------------------------------------------------------

class TestFretboardSVGNote:
    def test_shows_positions(self):
        g = Fretboard.violao()
        svg = FretboardSVG.note(g, Note("C"))
        assert "<svg" in svg
        assert ">C<" in svg


# ---------------------------------------------------------------------------
# Positions
# ---------------------------------------------------------------------------

class TestFretboardSVGPositions:
    def test_custom_title(self):
        g = Fretboard.violao()
        pos = g.positions(Note("E"))
        svg = FretboardSVG.positions(g, pos, "All E notes")
        assert "All E notes" in svg


# ---------------------------------------------------------------------------
# Field composite
# ---------------------------------------------------------------------------

class TestFretboardSVGField:
    def test_vertical_generates_svg(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.field(g, f)
        assert "<svg" in svg
        assert "</svg>" in svg

    def test_has_degree_labels(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.field(g, f)
        assert "I - CM" in svg
        assert "VII - Bdim" in svg

    def test_grid_layout(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.field(g, f, Layout.Grid)
        assert "<svg" in svg
        assert "I - CM" in svg

    def test_horizontal_layout(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.field(g, f, Layout.Horizontal)
        assert "<svg" in svg


# ---------------------------------------------------------------------------
# Progression composite
# ---------------------------------------------------------------------------

class TestFretboardSVGProgression:
    def test_generates_svg(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.progression(g, f, ["I", "IV", "V", "I"])
        assert "<svg" in svg

    def test_shows_labels(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.progression(g, f, ["I", "V"])
        assert "I (CM)" in svg
        assert "V (GM)" in svg


# ---------------------------------------------------------------------------
# Orientation
# ---------------------------------------------------------------------------

class TestFretboardSVGOrientation:
    def test_chord_horizontal(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"), 0, Orientation.Horizontal)
        assert "<svg" in svg
        assert "E4" in svg

    def test_scale_vertical(self):
        g = Fretboard.violao()
        svg = FretboardSVG.scale(g, Scale("C", ScaleType.Major), 0, 5,
                                  Orientation.Vertical)
        assert "<svg" in svg
        assert "Ionian" in svg

    def test_note_vertical(self):
        g = Fretboard.violao()
        svg = FretboardSVG.note(g, Note("C"), Orientation.Vertical)
        assert "<svg" in svg
        assert ">C<" in svg


# ---------------------------------------------------------------------------
# Handedness
# ---------------------------------------------------------------------------

class TestFretboardSVGHandedness:
    def test_chord_left_handed_differs(self):
        g = Fretboard.violao()
        svg_r = FretboardSVG.chord(g, Chord("CM"), 0,
                    Orientation.Vertical, Handedness.RightHanded)
        svg_l = FretboardSVG.chord(g, Chord("CM"), 0,
                    Orientation.Vertical, Handedness.LeftHanded)
        assert "<svg" in svg_r
        assert "<svg" in svg_l
        assert svg_r != svg_l

    def test_scale_left_handed_differs(self):
        g = Fretboard.violao()
        svg_r = FretboardSVG.scale(g, Scale("C", ScaleType.Major), 0, 5,
                    Orientation.Horizontal, Handedness.RightHanded)
        svg_l = FretboardSVG.scale(g, Scale("C", ScaleType.Major), 0, 5,
                    Orientation.Horizontal, Handedness.LeftHanded)
        assert "<svg" in svg_r
        assert "<svg" in svg_l
        assert svg_r != svg_l

    def test_chord_horizontal_left_handed(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("Am"), 0,
                  Orientation.Horizontal, Handedness.LeftHanded)
        assert "<svg" in svg
        assert "</svg>" in svg


# ---------------------------------------------------------------------------
# Composite with orientation/handedness
# ---------------------------------------------------------------------------

class TestFretboardSVGCompositeParams:
    def test_field_horizontal(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.field(g, f, Layout.Vertical, Orientation.Horizontal)
        assert "<svg" in svg
        assert "E4" in svg

    def test_progression_left_handed(self):
        g = Fretboard.violao()
        f = Field("C", "major")
        svg = FretboardSVG.progression(g, f, ["I", "V"],
                  Layout.Vertical, Orientation.Vertical, Handedness.LeftHanded)
        assert "<svg" in svg
        assert "I (CM)" in svg


# ---------------------------------------------------------------------------
# Default backward compatibility
# ---------------------------------------------------------------------------

class TestFretboardSVGBackwardCompat:
    def test_defaults_match_explicit(self):
        g = Fretboard.violao()
        svg1 = FretboardSVG.chord(g, Chord("CM"))
        svg2 = FretboardSVG.chord(g, Chord("CM"), 0,
                   Orientation.Vertical, Handedness.RightHanded)
        assert svg1 == svg2


# ---------------------------------------------------------------------------
# Full fretboard
# ---------------------------------------------------------------------------

class TestFretboardSVGFull:
    def test_generates_svg(self):
        g = Fretboard.violao()
        svg = FretboardSVG.full(g)
        assert "<svg" in svg
        assert "</svg>" in svg
        assert 'class="nut"' in svg

    def test_shows_instrument_name(self):
        g = Fretboard.violao()
        svg = FretboardSVG.full(g)
        assert g.name() in svg

    def test_vertical(self):
        g = Fretboard.violao()
        svg = FretboardSVG.full(g, Orientation.Vertical)
        assert "<svg" in svg


# ---------------------------------------------------------------------------
# Write
# ---------------------------------------------------------------------------

class TestFretboardSVGWrite:
    def test_writes_file(self):
        g = Fretboard.violao()
        svg = FretboardSVG.chord(g, Chord("CM"))
        with tempfile.TemporaryDirectory() as tmpdir:
            path = os.path.join(tmpdir, "test.svg")
            FretboardSVG.write(svg, path)
            assert os.path.exists(path)
            with open(path) as f:
                content = f.read()
            assert content == svg

File: tests/cpp/test_fretboard_svg.cpp
// Gingo — Music Theory Library
// Tests for FretboardSVG class.
//
// SPDX-License-Identifier: MIT

#include <catch2/catch_test_macros.hpp>
#include <gingo/fretboard_svg.hpp>
#include <gingo/field.hpp>

#include <filesystem>
#include <fstream>
#include <string>

using namespace gingo;

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

static bool contains(const std::string& haystack, const std::string& needle) {
    return haystack.find(needle) != std::string::npos;
}

// ---------------------------------------------------------------------------
// Chord diagram
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG chord generates valid SVG", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"));
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "</svg>"));
    REQUIRE(contains(svg, "xmlns=\"http://www.w3.org/2000/svg\""));
}

TEST_CASE("FretboardSVG chord contains title", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("Am7"));
    REQUIRE(contains(svg, "Am7"));
}

TEST_CASE("FretboardSVG chord has CSS style", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"));
    REQUIRE(contains(svg, "<style>"));
    REQUIRE(contains(svg, "</style>"));
}

TEST_CASE("FretboardSVG chord has string lines", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"));
    REQUIRE(contains(svg, "class=\"str\""));
}

TEST_CASE("FretboardSVG chord has fret lines", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"));
    REQUIRE(contains(svg, "class=\"frt\""));
}

TEST_CASE("FretboardSVG chord has dots", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"));
    REQUIRE(contains(svg, "class=\"dot\""));
}

TEST_CASE("FretboardSVG chord on cavaquinho", "[fretboard_svg]") {
    auto c = Fretboard::cavaquinho();
    auto svg = FretboardSVG::chord(c, Chord("CM"));
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "CM"));
}

TEST_CASE("FretboardSVG chord on bandolim", "[fretboard_svg]") {
    auto b = Fretboard::bandolim();
    auto svg = FretboardSVG::chord(b, Chord("CM"));
    REQUIRE(contains(svg, "<svg"));
}

// ---------------------------------------------------------------------------
// Fingering
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG fingering renders specific fingering", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto f = g.fingering(Chord("Am"));
    auto svg = FretboardSVG::fingering(g, f);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "Am"));
}

// ---------------------------------------------------------------------------
// Scale
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG scale generates valid SVG", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::scale(g, Scale("C", ScaleType::Major), 0, 12);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "</svg>"));
}

TEST_CASE("FretboardSVG scale contains mode name", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::scale(g, Scale("C", ScaleType::Major), 0, 5);
    REQUIRE(contains(svg, "Ionian"));
}

TEST_CASE("FretboardSVG scale has string labels", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::scale(g, Scale("C", ScaleType::Major), 0, 5);
    // Should contain open string labels like E4, B3, etc.
    REQUIRE(contains(svg, "E4"));
}

// ---------------------------------------------------------------------------
// Note
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG note shows all positions", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::note(g, Note("C"));
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, ">C<"));
}

// ---------------------------------------------------------------------------
// Positions
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG positions with custom title", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto pos = g.positions(Note("E"));
    auto svg = FretboardSVG::positions(g, pos, "All E notes");
    REQUIRE(contains(svg, "All E notes"));
}

// ---------------------------------------------------------------------------
// Field composite
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG field vertical generates SVG", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::field(g, f);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "</svg>"));
}

TEST_CASE("FretboardSVG field has degree labels", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::field(g, f);
    REQUIRE(contains(svg, "I - CM"));
    REQUIRE(contains(svg, "VII - Bdim"));
}

TEST_CASE("FretboardSVG field grid layout", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::field(g, f, Layout::Grid);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "I - CM"));
}

TEST_CASE("FretboardSVG field horizontal layout", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::field(g, f, Layout::Horizontal);
    REQUIRE(contains(svg, "<svg"));
}

// ---------------------------------------------------------------------------
// Progression composite
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG progression generates SVG", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::progression(g, f, {"I", "IV", "V", "I"});
    REQUIRE(contains(svg, "<svg"));
}

TEST_CASE("FretboardSVG progression shows labels", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::progression(g, f, {"I", "V"});
    REQUIRE(contains(svg, "I (CM)"));
    REQUIRE(contains(svg, "V (GM)"));
}

// ---------------------------------------------------------------------------
// Write
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Orientation
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG chord horizontal orientation", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"), 0, Orientation::Horizontal);
    REQUIRE(contains(svg, "<svg"));
    // Horizontal fretboard uses slbl class for string labels.
    REQUIRE(contains(svg, "E4"));
}

TEST_CASE("FretboardSVG scale vertical orientation", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::scale(g, Scale("C", ScaleType::Major), 0, 5,
                                     Orientation::Vertical);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "Ionian"));
}

TEST_CASE("FretboardSVG note vertical orientation", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::note(g, Note("C"), Orientation::Vertical);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, ">C<"));
}

// ---------------------------------------------------------------------------
// Handedness
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG chord left-handed", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg_r = FretboardSVG::chord(g, Chord("CM"), 0,
                     Orientation::Vertical, Handedness::RightHanded);
    auto svg_l = FretboardSVG::chord(g, Chord("CM"), 0,
                     Orientation::Vertical, Handedness::LeftHanded);
    // Both are valid SVGs.
    REQUIRE(contains(svg_r, "<svg"));
    REQUIRE(contains(svg_l, "<svg"));
    // They should differ (mirrored strings).
    REQUIRE(svg_r != svg_l);
}

TEST_CASE("FretboardSVG scale left-handed", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg_r = FretboardSVG::scale(g, Scale("C", ScaleType::Major), 0, 5,
                     Orientation::Horizontal, Handedness::RightHanded);
    auto svg_l = FretboardSVG::scale(g, Scale("C", ScaleType::Major), 0, 5,
                     Orientation::Horizontal, Handedness::LeftHanded);
    REQUIRE(contains(svg_r, "<svg"));
    REQUIRE(contains(svg_l, "<svg"));
    REQUIRE(svg_r != svg_l);
}

TEST_CASE("FretboardSVG chord horizontal left-handed", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("Am"), 0,
                   Orientation::Horizontal, Handedness::LeftHanded);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "</svg>"));
}

// ---------------------------------------------------------------------------
// Composite with orientation/handedness
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG field horizontal orientation", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::field(g, f, Layout::Vertical,
                                     Orientation::Horizontal);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "E4"));
}

TEST_CASE("FretboardSVG progression left-handed", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    Field f("C", ScaleType::Major);
    auto svg = FretboardSVG::progression(g, f, {"I", "V"},
                   Layout::Vertical, Orientation::Vertical,
                   Handedness::LeftHanded);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "I (CM)"));
}

// ---------------------------------------------------------------------------
// Default backward compatibility
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG defaults produce same output", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    // chord() defaults: Vertical, RightHanded — same as explicit.
    auto svg1 = FretboardSVG::chord(g, Chord("CM"));
    auto svg2 = FretboardSVG::chord(g, Chord("CM"), 0,
                    Orientation::Vertical, Handedness::RightHanded);
    REQUIRE(svg1 == svg2);
}

// ---------------------------------------------------------------------------
// Full fretboard
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG full generates SVG", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::full(g);
    REQUIRE(contains(svg, "<svg"));
    REQUIRE(contains(svg, "</svg>"));
    REQUIRE(contains(svg, "class=\"nut\""));
}

TEST_CASE("FretboardSVG full shows instrument name", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::full(g);
    REQUIRE(contains(svg, g.name()));
}

TEST_CASE("FretboardSVG full vertical", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::full(g, Orientation::Vertical);
    REQUIRE(contains(svg, "<svg"));
}

// ---------------------------------------------------------------------------
// Write
// ---------------------------------------------------------------------------

TEST_CASE("FretboardSVG write creates file", "[fretboard_svg]") {
    auto g = Fretboard::violao();
    auto svg = FretboardSVG::chord(g, Chord("CM"));
    std::string path = (std::filesystem::temp_directory_path() / "gingo_test_fretboard_svg.svg").string();
    FretboardSVG::write(svg, path);

    std::ifstream in(path);
    REQUIRE(in.is_open());
    std::string content((std::istreambuf_iterator<char>(in)),
                         std::istreambuf_iterator<char>());
    REQUIRE(content == svg);
    std::remove(path.c_str());
}
